### ISSUE TITLE
关于AdaRNN实现的一些问题

### ISSUE THREAD

**Issue by Arcueid-Brunestud** [jindongwang/transferlearning#318]
**Before you open an issue**
If it's code running error, maybe you want to check the python or pytorch version before submitting an issue.

**Describe the bug**

**To Reproduce**
Where is this bug happen?

**Screenshots**

**Comment by mengyays** [comment]
您好，我是杨允新，您的邮件我已收到，我会及时给您回复。

**Comment by Arcueid-Brunestud** [comment]
您好，想请问下一些关于AdaRNN的问题：

1. 数据的标准化是在用TDC划分前做的吗？
2. 目前代码的TDC部分是针对一个station的，想请问下如果是多个股票（论文中提到的一个例子）的话，TDC部分会有什么不同吗？还是直接比较不同domain内所有股票的分布？不同的domain内的股票也可能不同吧？
3. 代码是在测试集训练，验证集验证后，直接把训练集得到的模型作为最终模型，如果想要把训练验证合在一起后重新训练，那么TDC部分需要重新做一遍？超参仍然选2（似乎加上验证集后，可能会对domian的划分有影响吧）？其他参数也都使用验证集得到的参数？
4. 代码使用的数据集数据量比较小，您有尝试过对更大的数据量（千万以上），或者更多的特征时（训练alpha时会有一个很大的全连接网络）做优化吗？

希望能得到解答，谢谢！

**Comment by jindongwang** [comment]
1. Cosine loss在代码里没用上，只是在init里面随便写了一个默认的loss是cosine。实际中用的都是MMD和ADV，所以对结果没影响。然后确实这个cosine loss实现有问题，应该是.mean(0)。感谢提醒！（多亏我们没用这个loss）
2. 数据标准化在什么阶段都可以做，这个没有确定说法。
3. 多个station也可以强行把station的标签抹掉，当它是一个station，意思一样的。因为本质上AdaRNN就是要把一个大数据里面的隐含分布找出来，所以多个station也可以做。AdaRNN有一个升级版在这里，比AdaRNN更好，是端到端的：https://openreview.net/forum?id=NX0nX7TE4lc
4. 代码没考虑你说的这么多情况。就是训练集训练、验证集调参、测试集测试。训练验证合在一起训练这种操作应该是不合理的。
5. 没尝试过别的数据，你有兴趣可以尝试。数据大了、网络复杂了，肯定会慢。
6. 将alpha设置为等权重当然是有实验的，这是很重要的问题，我们当然有验证，具体可以看paper中的Figure 3(d)图。

**Comment by Arcueid-Brunestud** [comment]
> 1. Cosine loss在代码里没用上，只是在init里面随便写了一个默认的loss是cosine。实际中用的都是MMD和ADV（参照readme），所以对结果没影响。然后确实这个cosine loss实现有问题，应该是.mean(0)。感谢提醒！（多亏我们没用这个loss）
> 2. 数据标准化在什么阶段都可以做，这个没有确定说法。
> 3. 多个station也可以强行把station的标签抹掉，当它是一个station，意思一样的。因为本质上AdaRNN就是要把一个大数据里面的隐含分布找出来，所以多个station也可以做。AdaRNN有一个升级版在这里，比AdaRNN更好，是端到端的：https://openreview.net/forum?id=NX0nX7TE4lc
> 4. 代码没考虑你说的这么多情况。就是训练集训练、验证集调参、测试集测试。训练验证合在一起训练这种操作应该是不合理的。
> 5. 没尝试过别的数据，你有兴趣可以尝试。数据大了、网络复杂了，肯定会慢。
> 6. 将alpha设置为等权重当然是有实验的，这是很重要的问题，我们当然有验证，具体可以看paper中的Figure 3(d)图。

非常感谢您的解答，我还是有些疑问。
关于1，如果直接使用github中给出的示例命令运行应该就是使用cosine loss吧?
关于4，我想表达的意思是，时间序列的预测中一般训练、验证、测试按照时间顺序划分，所以如果预测时使用的模型没有用到最近的验证集数据（指训练模型）的话，一般预测会差很多，所以想问下该如何利用最近的这部分数据，这种情况下应该如何TDC。
关于6，我看了论文后没有搞得很清楚，如果我没有理解错的话，论文中分别有三种基本做法，pre-train代表着训练gru参数，不对权重做改进，naive代表着使用全连接网络学习权重，而boosting代表根据和上一轮结果的对比对权重调整。
然后，论文中是代表着figure 3(d)中是比较了naive/boosting with/without pre-train的四种情况，似乎没有始终等权的情况？
以及，代码中（train_adarnn）中的forward_pretrain中使用了全连接网络学习参数？boosting阶段是和论文中描述一致的。

另外，关于您最新的端到端论文，请问下目前有相关的代码吗，谢谢！

**Comment by jindongwang** [comment]
1. 这确实是我们的疏忽。我仔细看了一下，AdaRNN的readme后来被修改了（加入TDC后），我明天联系一下一作让他改过来。正常paper中对应最好结果应该是用了MMD和ADV loss（对应paper中Figure 5，主要对比就是这两种）。
4. 你说的没错，是按时间顺序划分的。如何利用，我们也没好的方法，就是把它当验证集，因为它毕竟不能参与训练。但是切分时间窗口构造训练数据时其实就有顺序信息（之前数据预测未来数据）。貌似也没有其他好的方法，这也是时间序列预测为什么这么难的原因。
6. 你可能误会了，c和d图的naive不是一个naive。如果alpha权重都是1，就对应的是不加alpha的情况，这在d图中被叫为naive（具体描述在section 5.7上一段"To evaluate the importance vector  alpha, we compare the results of with or without importance vector in Figure 3(d)."）这个naive就指的是权重为1的情况；而3c的naive是指用网络去学的情况。另外，我看了一下，figure 4b也进行了距离的展示，说明加入alpha机制分布距离更小。

**Comment by jindongwang** [comment]
最新那个paper你也看到了，被AC强拒了，尽管866的分数是中高分。因此在下一次中稿之前不会开源。

**Comment by jindongwang** [comment]
@yuntaodu

**Comment by yuntaodu** [comment]
感谢提醒，代码已经进行了修改。
